cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsConsumerMetaTable.java

```diff
@@ -65,20 +63,27 @@ public MetricsConsumerMetaTable(MetricsTable metaTable) {
 
 
   public <T extends MetricsMetaKey>
-  void saveMetricsProcessorStats(Map<T, MetricsProcessorStats> messageIds) throws Exception {
+  void saveMetricsProcessorStats(Map<T, TopicProcessMeta> messageIds) throws Exception {
     SortedMap<byte[], SortedMap<byte[], Long>> timestampUpdates = new TreeMap<>(Bytes.BYTES_COMPARATOR);
-    for (Map.Entry<T, MetricsProcessorStats> entry : messageIds.entrySet()) {
-      SortedMap<byte[], Long> timeMap = new TreeMap<>(Bytes.BYTES_COMPARATOR);
-      MetricsProcessorStats metaInfo = entry.getValue();
+    SortedMap<byte[], SortedMap<byte[], byte[]>> messageIdUpdates = new TreeMap<>(Bytes.BYTES_COMPARATOR);
+
+    for (Map.Entry<T, TopicProcessMeta> entry : messageIds.entrySet()) {
+      TopicProcessMeta metaInfo = entry.getValue();
       if (metaInfo.getMessagesProcessed() > 0L) {
+        SortedMap<byte[], Long> timeMap = new TreeMap<>(Bytes.BYTES_COMPARATOR);
         timeMap.put(PROCESS_COUNT_TOTAL, metaInfo.getMessagesProcessed());
         timeMap.put(PROCESS_TIMESTAMP_LATEST, metaInfo.getLatestMetricsTimestamp());
         timeMap.put(PROCESS_TIMESTAMP_OLDEST, metaInfo.getOldestMetricsTimestamp());
-        timeMap.put(PROCESS_TMS_PUBLISH_TIME, metaInfo.getPublishTimestamp());
+        timeMap.put(LAST_PROCESS_TIMESTAMP, metaInfo.getLastProcessedTimestamp());
         timestampUpdates.put(entry.getKey().getKey(), timeMap);
+
+        SortedMap<byte[], byte[]> messageIdMap = new TreeMap<>(Bytes.BYTES_COMPARATOR);
+        messageIdMap.put(MESSAGE_ID_COLUMN, entry.getValue().getMessageId());
+        messageIdUpdates.put(entry.getKey().getKey(), messageIdMap);
       }
     }
     metaTable.put(timestampUpdates);
+    metaTable.putBytes(messageIdUpdates);
```

Comment:
can you do it in a single put? That makes it atomic - HBase will either write the row or not.